bitcoin/src/bip32.rs

```diff
@@ -369,107 +646,238 @@ impl<'a> Iterator for DerivationPathIterator<'a> {
     fn next(&mut self) -> Option<Self::Item> {
         let ret = self.next_child?;
         self.next_child = ret.increment().ok();
-        Some(self.base.child(ret))
+        Some(self.base.with_child(ret))
     }
 }
+/// [`DerivationPath`] methods and utils.
+pub mod derivation_path {
 
-impl DerivationPath {
-    /// Returns length of the derivation path
-    pub fn len(&self) -> usize { self.0.len() }
+    use super::*;
+    /// A BIP 32 derivation path.
+    #[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
+    pub struct DerivationPath(Vec<ChildNumber>);
+    #[cfg(feature = "serde")]
+    internals::serde_string_impl!(DerivationPath, "a BIP-32 derivation path");
+    impl DerivationPath {
+        /// The master(empty) path (i.e. the master key).
+        pub const MASTER: Self = DerivationPath(Vec::new());
+
+        /// Create a new [`DerivationPath`] from a vector of [`ChildNumber`].
+        pub fn new(vec: Vec<ChildNumber>) -> Self { DerivationPath(vec) }
+        /// Try to convert this [`DerivationPath`] into a [`NormalDerivationPath`]
+        pub fn to_normal(&self) -> Result<NormalDerivationPath, Error> {
+            NormalDerivationPath::try_from(self.0.as_slice())
+        }
+        /// Check if the [`DerivationPath`] is empty
+        pub fn is_empty(&self) -> bool { self.0.is_empty() }
+        /// Returns length of the [`DerivationPath`]
+        pub fn len(&self) -> usize { self.0.len() }
+
+        /// Returns whether derivation path represents master key (i.e. it's length
+        /// is empty). True for `m` path.
+        pub fn is_master(&self) -> bool { self.0.is_empty() }
+
+        /// Create a new [`DerivationPath`] that is a child of this one.
+        pub fn with_child(&self, cn: ChildNumber) -> Self { self.new_extended_from([cn]) }
+
+        /// Convert into a [`DerivationPath`] that is a child of this one.
+        pub fn into_with_child(self, cn: ChildNumber) -> Self {
+            let mut ret = self.clone();
+            ret.extend_from([cn]);
+            ret
+        }
 
-    /// Returns `true` if the derivation path is empty
-    pub fn is_empty(&self) -> bool { self.0.is_empty() }
+        /// Get an [`Iterator`] over the children of this [`DerivationPath`]
```

Comment:
```suggestion
        /// Returns an [`Iterator`] over the children of this [`DerivationPath`]
```
You are not getting anything. Get implies that something pre-exists.